duckduckgo/Android
# Create Dax Dialog component

Task/Issue URL: https://app.asana.com/0/1125189844152671/1146799866806650
Tech Design URL: 
https://app.asana.com/0/1125189844152671/1148078837309948
https://app.asana.com/0/1125189844152671/1148078837309945
https://app.asana.com/0/1125189844152671/1148078837309949

**Description**:
To develop a reusable component which follows the specification on https://app.asana.com/0/1125189844152671/1146799866806650

**Steps to test this PR**:
The dialog exists as its own component and is not used anywhere yet so you will have to create a new instance of it. To do that you can add the following code inside the `renderLoadingIndicator` function in `BrowserTabFragment`. 

```
if (viewState.progress == 100 ) {
    val fireButton = requireActivity().findViewById<View>(R.id.fire)
    val gradeView = requireActivity().findViewById<View>(R.id.privacyGradeButton)

    val text = "Every site has a privacy grade. Tap it to see how I protected your privacy.\n\nWant to get fancy? Try clearing your data by hitting the fire button."
     val dialog = DaxDialog(text, "Got it")

     dialog.onAnimationFinishedListener {
         dialog.startHighlightViewAnimation(fireButton, timesBigger = 0.2f)
         dialog.startHighlightViewAnimation(targetView = gradeView, timesBigger = 0.7f)
     }
     dialog.show(requireFragmentManager(), "test")
}
```

**Dialog typing animation**
1. Launch the app using the code above and navigate to any website.
1. The dialog should show and a typing animation start to run.

**Dialog animation finished listener***
1. Launch the app using the code above and navigate to any website.
1. After the typing animation finishes the code added in `dialog.onAnimationFinishedListener` should be executed.

**Dialog view highlighted**
1. Launch the app using the code above and navigate to any website.
1. After the typing animation finished the fire button and the grade button should be highlighted within the dialog.

**Dialog view highlighted change size**
1. Change the code in the listener to be `dialog.startHighlightViewAnimation(targetView = gradeView, timesBigger = 2f)`
1. Launch the app using the code above and navigate to any website.
1. After the typing animation finishes note how the circle highlighting the grade button is smaller.

**Cancel typing animation**
1. Launch the app using the code above and navigate to any website.
1. Before the typing animation finishes tap in the text view to cancel the animation.
1. The whole text should be displayed and the onAnimationFinishedListener executed.

**Dismiss the dialog tapping in the background**
1. Launch the app using the code above and navigate to any website.
1. Tapping in the background at any point should dismiss the dialog even if the typing animation is running in which case it will get cancelled accordingly.

**Dismiss the dialog tapping in the hide button**
1. Launch the app using the code above and navigate to any website.
1. Tapping in the hide button at any point should dismiss the dialog even if the typing animation is running in which case it will get cancelled accordingly.

**Make dialog not dismissible**
1. Change the code above to `val dialog = DaxDialog(text, "Got it", false)`.
1. Launch the app using the code above and navigate to any website. 
1. Tapping in the background should not dismiss the dialog however tapping the hide button will.

**Change speed of the typing animation**
1. Change the code above to `val dialog = DaxDialog(text, "Got it", true, 100)`.
1. Launch the app using the code above and navigate to any website. 
1. Notice how increase the typing animation delay from 20ms to 100ms slows the typing animation down.

**Change listeners**
1. Feel free to set `setPrimaryCTAClickListener` and `setHideClickListener` to override the behaviour of the primary CTA button and the hide button. 
1. Notice how the animations are still cancelled to avoid issues in case the developer forgets about that.
1. Dismissing the dialog is not done for the developer so will have to be manually set while creating the listener.
---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
